src/librustdoc/clean/finder_trait.rs

```diff
@@ -0,0 +1,154 @@
+// Copyright 2018 The Rust Project Developers. See the COPYRIGHT
+// file at the top-level directory of this distribution and at
+// http://rust-lang.org/COPYRIGHT.
+//
+// Licensed under the Apache License, Version 2.0 <LICENSE-APACHE or
+// http://www.apache.org/licenses/LICENSE-2.0> or the MIT license
+// <LICENSE-MIT or http://opensource.org/licenses/MIT>, at your
+// option. This file may not be copied, modified, or distributed
+// except according to those terms.
+
+use rustc::hir;
+use syntax_pos::DUMMY_SP;
+
+use super::*;
+
+pub trait Finder<'a, 'tcx: 'a, 'rcx: 'a> {
+    fn get_cx(&self) -> &DocContext<'a, 'tcx, 'rcx>;
+
+    // This is an ugly hack, but it's the simplest way to handle synthetic impls without greatly
+    // refactoring either librustdoc or librustc. In particular, allowing new DefIds to be
+    // registered after the AST is constructed would require storing the defid mapping in a
+    // RefCell, decreasing the performance for normal compilation for very little gain.
+    //
+    // Instead, we construct 'fake' def ids, which start immediately after the last DefId in
+    // DefIndexAddressSpace::Low. In the Debug impl for clean::Item, we explicitly check for fake
+    // def ids, as we'll end up with a panic if we use the DefId Debug impl for fake DefIds
+    fn next_def_id(&self, crate_num: CrateNum) -> DefId {
```

Comment:
Why would you need new `DefId`s?!

Reply:
To avoid them getting filtered out because of duplicates.

Reply:
Ah hold on, it's required for the auto traits (sync and send) but not be for blanket impls.

Reply:
Still, *why*?

Reply:
No it's required for blanket impls as well otherwise they're counted as duplicates.

Reply:
You should change the deduplication logic instead, then!

Reply:
Yep but nothing to do with this PR. :)

Reply:
Every `clean::Item` needs a `DefId`. As the synthetic impls for `auto` traits (and hopefully eventually, blanket traits too) invent an impl out of thin air, one needs to be created so that the `Item` can hold it.

The real question is whether we can get away with only creating `clean::Impl` structs, and ignoring/inferring the fields on `Item`, like its attributes, visibility, stability, or deprecation. These can probably be inferred from the trait itself (or the blanket impl being applied), so that would help to prevent us from requiring new `DefId`s.